### PR TITLE
Add support for an alternate workspaceRoot.

### PR DESCRIPTION
**Pull request by indierojo**
Adds an optional configuration option to change the path used to find the testcafe binary. This is useful for cases where the test files are kept in a nested directory independent of the source files they are testing but contained within the same repository.

```
| {workspaceRoot}
| node_modules <-- no testcafe
| src\
\ acceptance\
  |- node_modules <-- testcafe
```

`default`: would try to find testcafe in the root/node_modules,
`"testcafeTestRunner.workspaceRoot": "/acceptance"`: would try to find testcafe in root/acceptance/node_modules

### CODE REVIEW

**Comment by romanresh** [comment]
Hi @indierojo. It's unclear to me, why you need to place the node_module folder with testcafe in a subfolder. Do you have a separate package.json file inside the subfolder? How does it work?

FYI: I had an idea to support the globally installed testcafe (for instance, via setting the absolute path to testcafe in the configuration file) but your task looks a bit different.

**Comment by indierojo** [comment]
The use case is definitely a bit non-standard. I do have a separate package.json file for a couple reasons:
1. I publish them to a private repository independently of the main repository
2. so I can have fewer dependencies that I need to install in order to run the tests (ie: in CI)

At the same time I like to keep them co-located with the files being tested, hence not putting them in a repository of their own.

I think the global testcafe route would also work to solve my use case so I'm open to that if this seems too niche to merge.

**Comment by indierojo** [comment]
Also, with tools like Lerna and now Yarn workspaces, there are probably a few other folks out there who may run into this problem.

**Comment by romanresh** [comment]
Ok, it makes sense.

File: src/extension.ts [hunk 1]
```diff
@@ -244,7 +252,8 @@ class TestCafeTestController {
             args.push(name);
         }
 
-        var testCafePath = vscode.workspace.rootPath + TESTCAFE_PATH;
+        const workspacePathOverride = this.getOverriddenWorkspacePath()
+        var testCafePath = vscode.workspace.rootPath + workspacePathOverride + TESTCAFE_PATH;
```
Comment (romanresh) [inline src/extension.ts:256]: I believe, it's better to replace the strings concatenation with the path.join() function.

File: src/extension.ts [hunk 2]
```diff
@@ -254,9 +263,9 @@ class TestCafeTestController {
             "type": "node2",
             "request": "launch",
             "name": "Launch current test(s) with TestCafe",
-            "program": "${workspaceRoot}" + TESTCAFE_PATH,
+            "program": testCafePath,
             "args": args,
-            "cwd": "${workspaceRoot}",
+            "cwd": "${workspaceRoot}" + workspacePathOverride,
```
Comment (romanresh) [inline src/extension.ts:268]: Looks like the slash symbol is missed here.

**Review by romanresh — DISMISSED** [review]

**Comment by romanresh** [comment]
Hi @indierojo . Sorry for long pause in my answer, I was really overwhelmed at work this week. I'll merge your PR to a separate branch and add a few fixes. I think, it's better to use really relative paths in the settings (which can start from "./" or a folder name, without the '/' symbol). So, instead of "/tests" you will have to use the "./tests" setting value.

**Comment by indierojo** [comment]
Thanks for the merge - your changes look great to me -- I agree that relative is a better way to go.

**Pull request merged**